x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/analytics/process/AnalyticsProcessManager.java

```diff
@@ -0,0 +1,80 @@
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.ml.analytics.process;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper;
+import org.elasticsearch.xpack.ml.MachineLearning;
+import org.elasticsearch.xpack.ml.analytics.DataFrameDataExtractor;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ExecutorService;
+
+public class AnalyticsProcessManager {
+
+    private static final Logger LOGGER = LogManager.getLogger(AnalyticsProcessManager.class);
+
+    private final Client client;
+    private final Environment environment;
+    private final ThreadPool threadPool;
+    private final AnalyticsProcessFactory processFactory;
+
+    public AnalyticsProcessManager(Client client, Environment environment, ThreadPool threadPool,
+                                   AnalyticsProcessFactory analyticsProcessFactory) {
+        this.client = Objects.requireNonNull(client);
+        this.environment = Objects.requireNonNull(environment);
+        this.threadPool = Objects.requireNonNull(threadPool);
+        this.processFactory = Objects.requireNonNull(analyticsProcessFactory);
+    }
+
+    public void processData(String jobId, DataFrameDataExtractor dataExtractor) {
+        threadPool.generic().execute(() -> {
+            AnalyticsProcess process = createProcess(jobId);
+            try {
+                // Fake header
+                process.writeRecord(dataExtractor.getFieldNamesArray());
+
+                while (dataExtractor.hasNext()) {
+                    Optional<List<String[]>> records = dataExtractor.next();
+                    if (records.isPresent()) {
+                        for (String[] record : records.get()) {
+                            process.writeRecord(record);
+                        }
+                    }
+                }
+                process.flushStream();
+
+                LOGGER.debug("[{}] Closing process", jobId);
+                process.close();
+                LOGGER.info("[{}] Closed process", jobId);
+            } catch (IOException e) {
+
+            } finally {
+                try {
+                    process.close();
+                } catch (IOException e) {
+                    LOGGER.error("[{}] Error closing data frame analyzer process", jobId);
+                }
+            }
+        });
+    }
+
+    private AnalyticsProcess createProcess(String jobId) {
+        ExecutorService executorService = threadPool.executor(MachineLearning.AUTODETECT_THREAD_POOL_NAME);
```

Comment:
I think it's good to reuse this threadpool, but it implies that the analytics process is a "job" from the point of view of deciding how many jobs can run on each node.  We definitely need to hook these processes into a unified allocation framework.  Also, maybe rename the threadpool to `JOB_PROCESS_THREAD_POOL_NAME` or similar.

Reply:
I think we should have just one pool for all job types, and the maximum number of open jobs per node should also relate to _all_ job types.  Since the number of threads in the pool is governed by the number of named pipes we create per job we shouldn't need to adjust the thread pool size.  However, since people might now want more jobs this might be a good opportunity to implement #29809.